rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -3135,6 +3147,16 @@ extern "C" rmw_ret_t rmw_send_response(
   cdds_request_header_t header;
   memcpy(&header.guid, request_header->writer_guid, sizeof(header.guid));
   header.seq = request_header->sequence_number;
+  // if the number of writers matching our request reader equals the number
+  // of readers matching our response writer, we have a pretty decent claim
+  // to have discovered this client's response reader.
```

Comment:
@eboasson nit: consider moving this comment block closer to the `check_for_response_reader()` function.

Reply:
> we have a pretty decent claim

:eyes:

Reply:
Any chance we can tell if remote writer and reader belong to the same server instance? Or at least to the same participant?

Reply:
I actually started working out how to do that, and it is definitely possible the determine they are from the same participant. But now that all nodes in a process (well, context) share the same participant that seems insufficiently selective. I couldn’t find a way to tie them to the node (which I think is selective enough) short of using `USER_DATA`. That seemed like it might be unwise. I suppose the graph cache interface could be extended to answer this question.

What we do know here is that we have discovered the client’s writer (else we could never have received its request), and that therefore the client’s writer is accounted for in the “subscription matched” status. Thus, it is pretty likely that the service writer’s match count won’t equal the service’s reader match count unless that client’s reader has been discovered. But there is no guarantee: if another client shows up and its reader happens to be discovered before its writer, you will draw the wrong conclusion.

In the absence of packet loss, that is highly unlikely because it creates the writer first, because under that assumption, the discovery of the writer will precede the discovery of the reader. Indeed, I wouldn’t be surprised if simply swapping the creation of the reader and the writer in the client would make the problem disappear in the test setup.

Reply:
> Indeed, I wouldn’t be surprised if simply swapping the creation of the reader and the writer in the client would make the problem disappear in the test setup.

Interesting. Have you tried it?